### PR TITLE
Suggest replacing named lifetime with `'static`

### PR DESCRIPTION
**Pull request by estebank**
```
warning: unnecessary lifetime parameter `'a`
  --> file.rs:11:14
   |
11 | fn foo<'c, 'a: 'static, 'b, 'd>(_x: &'a str) -> &'a str {
   |            ^^^^^^^^^^^
help: you can use the `'static` lifetime directly, in place of `'a`
   |
11 | fn foo<'c, 'b, 'd>(_x: &'static str) -> &'static str {
   |           --            ^^^^^^^          ^^^^^^^

warning: unnecessary lifetime parameter `'a`
  --> file.rs:14:9
   |
14 | fn bar<'a: 'static>(_x: &'a str) -> &'a str {
   |        ^^^^^^^^^^^
help: you can use the `'static` lifetime directly, in place of `'a`
   |
14 | fn bar(_x: &'static str) -> &'static str {
   |      --     ^^^^^^^          ^^^^^^^
```

cc #36179

### CODE REVIEW

**Comment by estebank** [comment]
Doesn't handle lifetimes that are not directly in an argument nor the following case:

```
warning: unnecessary lifetime parameter `'a`
  --> file.rs:11:14
   |
11 | fn _foo3<'c, 'a: 'static, 'b: 'a, 'd>(_x: &'a str) -> &'a str {
   |              ^^^^^^^^^^^
help: you can use the `'static` lifetime directly, in place of `'a`
   |
11 | fn _foo3<'c, 'b: 'a, 'd>(_x: &'static str) -> &'static str {
   |             --                ^^^^^^^
```

**Comment by pietroalbini** [comment]
Ping from triage @nikomatsakis! This PR needs your review.

**Comment by estebank** [comment]
Ping.

**Comment by nikomatsakis** [comment]
@estebank I'm not entirely sure how this is related to https://github.com/rust-lang/rust/issues/36179 -- it seems like a distinct problem. Do we think that it's common to write `'a: 'static`? I do it sometimes as a kind of type-system hack, but I'm not sure if it is something that arises "in the wild".

**Comment by nikomatsakis** [comment]
Overall, this seems like a bit more code than I would have expected; it feels like there should be a lighterweight way to detect this condition.

**Comment by nikomatsakis** [comment]
Well, I guess that making the edit is part of the hard part...

**Comment by estebank** [comment]
This code can probably be cleaned quite a bit, but wanted to get some feedback on wether this is worthwhile to begin with.

This doesn't fix #36179, but by providing a suggestion for lifetimes the risk of someone writing `'&str` is much lower. I used this case in particular to start experimenting on how suggestions for changing multiple lifetimes would look like. If we're ok with the output (and the increased amount of code to be able to get and operate over the lifetime spans), this would only be the first of many other suggestions being reworked from pure text to structured suggestion.

**Comment by nikomatsakis** [comment]
I see. I think the output looks good. I'm not sure if there is a more light-weight way to get it.

**Comment by nikomatsakis** [comment]
I made one suggestion that would help to remove a lot of the boilerplate (using a visitor)

**Comment by estebank** [comment]
CC https://github.com/rust-lang/rust/issues/44506 (same mechanism can be used to suggest the proposed output).

File: src/librustc/middle/resolve_lifetime.rs [hunk 1]
```diff
@@ -1441,6 +1446,18 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
             .collect();
 
         let next_early_index = index + generics.ty_params().count() as u32;
+        let mut arg_lifetimes = vec![];
+        for input in &decl.inputs {
```
Comment (nikomatsakis) [inline src/librustc/middle/resolve_lifetime.rs]: This vector is only used in the case of lint/error, right? I'm not sure how I feel about building it eagerly -- maybe we can convert this into a closure or a trait or something?

File: src/librustc/middle/resolve_lifetime.rs [hunk 2]
```diff
@@ -2149,82 +2176,147 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
...
-                            ))
-                            .emit();
-                    }
-                    hir::LifetimeName::Fresh(_)
-                    | hir::LifetimeName::Implicit
-                    | hir::LifetimeName::Name(_) => {
-                        self.resolve_lifetime_ref(bound);
+                            );
+
+                            // all the use spans for this lifetime in arguments to be replaced
```
Comment (nikomatsakis) [inline src/librustc/middle/resolve_lifetime.rs]: we should extract this into a helper fn, but the basic idea seems pretty good 💯

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
OK, @estebank -- now that I understand better what you were going for, I think this looks pretty nice. It seems like we should (A) rewrite to use a visitor,  (B) make the vector construction lazy, and (C) add a helper fn, but otherwise this seems like a nice POC. (We could/should fix shadowing too but that is less urgent since it is currently illegal anyway)

**Comment by nikomatsakis** [comment]
As an example where shadowing could be relevant though:

```rust
fn foo<'a: 'static>(x: for<'a> fn(&'a u32)) { }
```

File: src/librustc/middle/resolve_lifetime.rs (hunk 1 above)
Comment (estebank) [reply src/librustc/middle/resolve_lifetime.rs]: Makes sense, I'll see what I can do about it.

File: src/librustc/hir/mod.rs [hunk 3]
```diff
@@ -1662,6 +1669,137 @@ pub struct Ty {
     pub hir_id: HirId,
 }
 
+impl Ty {
+    pub fn lifetimes(&self) -> Vec<Lifetime> {
```
[the remark this answers is not in the document]
Comment (estebank) [reply src/librustc/hir/mod.rs]: It's only used to gather all the lifetime spans in the argument and return tys that match the name being checked.

**Comment by estebank** [comment]
The output for 

```
fn foo<'a: 'static>(x: for<'a> fn(&'a u32)) { }
```

is

```
warning: unnecessary lifetime parameter `'a`
 --> file.rs:1:8
  |
1 | fn foo<'a: 'static>(x: for<'a> fn(&'a u32)) { }
  |        ^^^^^^^^^^^
help: you can use the `'static` lifetime directly, in place of `'a`
  |
1 | fn foo(x: for<'a> fn(&'static u32)) { }
  |      --               ^^^^^^^

error[E0496]: lifetime name `'a` shadows a lifetime name that is already in scope
 --> file.rs:1:28
  |
1 | fn foo<'a: 'static>(x: for<'a> fn(&'a u32)) { }
  |        --                  ^^ lifetime 'a already in scope
  |        |
  |        first declared here
```

**Comment by nikomatsakis** [comment]
@estebank 

> The output for...

hmm, that seems suboptimal, do you agree? I think it'd be pretty easy to fix -- in the `lifetimes()` function that gathers up the lifetimes that appear in a type, when we encounter types with `for<'a>` binders, we should remove the things that appear in the binders from the result. (i.e., we only want the "free" lifetimes of the type)

**Comment by estebank** [comment]
Agree with your comments. I tried rebasing to see how hard it would be to include the feedback and the codebase diverged significantly. I'll get back to this later.

**Comment by nikomatsakis** [comment]
@estebank should we close the PR in the mean time?

**Pull request closed without merging**